bin/_test-helpers.sh

```diff
@@ -0,0 +1,318 @@
+        echo ""
+        echo "Usage:"
+        echo "    ${0##*/} [--images] [--images-host ssh://linkerd-docker] [--name test-name] [--skip-kind-create] /path/to/linkerd"
+        echo ""
+        echo "Examples:"
+        echo "    # Run all tests in isolated clusters"
+        echo "    ${0##*/} /path/to/linkerd"
+        echo ""
+        echo "    # Run single test in isolated clusters"
+        echo "    ${0##*/} --name test-name /path/to/linkerd"
+        echo ""
+        echo "    # Skip KinD cluster creation and run all tests in default cluster context"
+        echo "    ${0##*/} --skip-kind-create /path/to/linkerd"
+        echo ""
+        echo "    # Load images from tar files located under the 'image-archives' directory"
+        echo "    # Note: This is primarly for CI"
+        echo "    ${0##*/} --images /path/to/linkerd"
+        echo ""
+        echo "    # Retrieve images from a remote docker instance and then load them into KinD"
+        echo "    # Note: This is primarly for CI"
+        echo "    ${0##*/} --images --images-host ssh://linkerd-docker /path/to/linkerd"
+        echo ""
+        echo "Available Commands:"
+        echo "    --name: the argument to this option is the specific test to run"
+        echo "    --skip-kind-create: skip KinD cluster creation step and run tests in an existing cluster."
+        echo "    --images: (Primarily for CI) use 'kind load image-archive' to load the images from local .tar files in the current directory."
+        echo "    --images-host: (Primarily for CI) the argument to this option is used as the remote docker instance from which images are first retrieved (using 'docker save') to be then loaded into KinD. This command requires --images."
+        exit 0
+        ;;
+      --images)
+        images=1
+        ;;
+      --images-host)
+        images_host=$2
+        if [ -z "$images_host" ]; then
+          echo "Error: the argument for --images-host was not specified"
+          exit 1
+        fi
+        shift
+        ;;
+      --name)
+        test_name=$2
+        if [ -z "$test_name" ]; then
+          echo "Error: the argument for --name was not specified"
+          exit 1
+        fi
+        shift
+        ;;
+      --skip-kind-create)
+        skip_kind_create=1
+        ;;
+      *)
+        break
+    esac
+    shift
+  done
+
+  if [ "$images_host" ] && [ -z "$images" ]; then
+    echo "Error: --images-host needs to be used with --images" >&2
+    exit 1
+  fi
+
+  export linkerd_path="$1"
+  if [ -z "$linkerd_path" ]; then
+    echo "Error: path to linkerd binary is required"
+    echo "Help:"
+    echo "     ${0##*/} -h|--help"
+    echo "Basic usage:"
+    echo "     ${0##*/} /path/to/linkerd"
+    exit 64
+  fi
+}
+
+test_setup() {
+  bindir=$( cd "${BASH_SOURCE[0]%/*}" && pwd )
+  export bindir
+
+  export test_directory="$bindir"/../test
+
+  check_linkerd_binary
+}
+
+check_linkerd_binary(){
+  printf 'Checking the linkerd binary...'
```

Comment:
Using `echo` would be better, since it takes less resources than `printf` and no `printf`-specific functionality is used here.

Reply:
I'm going to keep this `printf`. We use the fact that it doesn't start a new line, which the `[ok]` relies on below so that it prints on the same line.

Reply:
And this is my same response for all the below suggestions about changing `printf` to`echo`!

Reply:
Ah, then `printf`-specific functionality _is_ indeed used. My bad. :)